game/libda/portaudio.cc

```diff
@@ -0,0 +1,120 @@
+				PaDeviceInfo const* info = Pa_GetDeviceInfo(Pa_HostApiDeviceIndexToDeviceIndex(backendIndex, i));
+				if (!info) continue;
+				std::string name = UnicodeUtil::convertToUTF8(info->name);
+				/// Omit some useless legacy devices of PortAudio/ALSA from our list
+				for (auto const& dev: { "front", "surround40", "surround41", "surround50", "surround51", "surround71", "iec958", "spdif", "dmix" }) {
+					if (name.find(dev) != std::string::npos) name.clear();
+				}
+				if (name.empty()) continue;  // No acceptable device found
+				devices.emplace_back(DeviceInfo(i, name, info->maxInputChannels, info->maxOutputChannels, Pa_HostApiDeviceIndexToDeviceIndex(backendIndex, i)));
+			}
+			for (auto& dev: devices) {
+				// Array of regex - replacement pairs
+				static char const* const replacements[][2] = {
+					{ "\\(hw:\\d+,", "(hw:," },  // Remove ALSA device numbers
+					{ " \\(.*\\)", "" },  // Remove the parenthesis part entirely
+				};
+				for (auto const& rep: replacements) {
+					std::string flex = std::regex_replace(dev.flex, std::regex(rep[0]), rep[1]);
+					if (flex == dev.flex) continue;  // Nothing changed
+									 // Verify that flex doesn't find any wrong devices
+					bool fail = false;
+					try {
+						if (find(flex, false, 0).idx != dev.idx) fail = true;
+					} catch (...) {}  // Failure to find anything is success
+					if (!fail) dev.flex = flex;
+				}
+			}
+		}
+		/// Get a printable dump of the devices
+		std::string dump() const {
+			std::ostringstream oss;
+			for (auto const& d: devices) { oss << "    #" << d.idx << " " << d.desc() << std::endl; }
+			return oss.str();
+		}
+		DeviceInfo const& find(std::string const& name, bool output, unsigned num) {
+			if (name.empty()) { return findByChannels(output, num); }
+			// Try name search with full match
+			for (auto const& dev: devices) {
+				if ((output ? dev.out : dev.in) < num) continue;
+				if (dev.name == name) return dev;
+			}
+			// Try name search with partial/flexible match
+			for (auto const& dev: devices) {
+				if ((output ? dev.out : dev.in) < num) continue;
+				if (dev.name.find(name) != std::string::npos) { return dev; }
+				if (dev.flex.find(name) != std::string::npos) { return dev; }
+			}
+			throw std::runtime_error("No such device.");
+		}
+		DeviceInfo const& findByChannels(bool output, unsigned num) {
+			for (auto const& dev: devices) {
+				unsigned reqChannels = output ? dev.out : dev.in;
+				if (reqChannels >= num) { return dev;  }
+			}
+			throw std::runtime_error("No such device.");
+		}
+		DeviceInfos devices;
+	};
+
+	struct AudioBackendFactory::Init {
+		Init() { PORTAUDIO_CHECKED(Pa_Initialize, ());
+			for (int i = 0; i < Pa_GetHostApiCount(); ++i) {
+				backends.emplace_back(Pa_GetHostApiInfo(i));
+			}
+			if (backends.empty()) throw std::runtime_error("No suitable audio backends found.");
+		}
+		~Init() { Pa_Terminate(); }
+		std::vector<const PaHostApiInfo *> backends;
```

Comment:
Are we certain PortAudio cleans this appropriately? If not, maybe use a smart pointer?

Reply:
The raw pointers are her on purpose, this is a non ownership case: PaHostApiInfo are provided and owned by portaudio and *must* not be deleted by user. For now, there is no smart pointer to indicate non ownership (discussions are still going on in the standard).

Reply:
Fair enough